src/main/java/io/jenkins/plugins/xygeni/saltbuildstep/SaltProvenanceRecorder.java

```diff
@@ -0,0 +1,308 @@
+        this.otherSubjectsOn = othersOn;
+        if (!this.otherSubjectsOn) {
+            this.subjects = null;
+        }
+    }
+
+    public boolean getOnOtherSubjects() {
+        return this.otherSubjectsOn;
+    }
+
+    public boolean isOtherSubjectsOn() {
+        return this.otherSubjectsOn;
+    }
+
+    // constructor
+    @DataBoundConstructor
+    public SaltProvenanceRecorder(String artifactFilter, List<Subject> subjects) {
+        this.artifactFilter = artifactFilter;
+        this.subjects = subjects;
+    }
+
+    public String getName() {
+        return COMMANDNAME;
+    }
+
+    @Override
+    public void perform(
+            @NonNull Run<?, ?> run,
+            @NonNull FilePath workspace,
+            @NonNull EnvVars env,
+            @NonNull Launcher launcher,
+            @NonNull TaskListener listener) throws IOException, InterruptedException {
+
+        PrintStream console = listener.getLogger();
+
+        if (run.getResult() != Result.SUCCESS) {
+            console.println("[xygeniSalt] - build not successful, abort generating provenance attestations");
+            return;
+        }
+
+        List<Subject> subjects = new ArrayList<>();
+
+        if (getArtifactFilter() != null) {
+            addSubjects(workspace, env, listener, getArtifactFilter(), subjects);
+        }
+
+        if (getSubjects() != null) {
+            subjects.addAll(getSubjects());
+        }
+
+        XygeniSaltCommand.run(
+                run, launcher, listener, getKey(), getKeyPassword(), getSubjects());
+    }
+
+    private void addSubjects(FilePath workspace, EnvVars env, TaskListener listener, String artifactFilter, List<Subject> subjects) throws IOException, InterruptedException {
+
+        PrintStream console = listener.getLogger();
+
+        String expandedFilter = env.expand(artifactFilter);
+        FilePath[] artifacts  = workspace.list(expandedFilter);
+
+        console.println("[xygeniSalt] collecting artifacts");
+
+        for (FilePath artifact : artifacts) {
+            console.println(" > " + artifact.getRemote());
+            subjects.add(Subject.of(artifact, workspace));
+        }
+    }
+
+    // Overridden for better type safety.
+    // If your plugin doesn't really define any property on Descriptor,
+    // you don't have to do this.
+    @Override
+    public DescriptorImpl getDescriptor() {
+        return (DescriptorImpl) super.getDescriptor();
+    }
+
+    /**
+     * Descriptor for {@link SaltProvenanceRecorder}.
+     */
+    @Symbol("xygeniSalt")
+    @Extension
+    public static final class DescriptorImpl extends BuildStepDescriptor<Publisher> {
+
+        // descritor
+
+        public static DescriptorImpl get() {
+            return ExtensionList.lookupSingleton(DescriptorImpl.class);
+        }
+
+        // CHECKS
+
+        @RequirePOST
+        public FormValidation doCheckKey(@QueryParameter String key) {
+            if (key.isEmpty()) {
```

Comment:
Probably a stronger validation is needed. See `io.xygeni.salt.sign.SignerKeyLoader.validate(...)` for further details. If the user gives the PEM keyvalue, validation is possible. I am not sure that when the user creates the build post-action using the UI, environment variables (for the `env:` prefix) or files (for the `file:` prefix) are available, as they are probably available for the runner running the job at runtime, so validation may not be exactly the same.

Reply:
added